extensions/amp-link-rewriter/0.1/link-rewriter.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {getConfigOpts} from './config-options';
+import {getDataParamsFromAttributes} from '../../../src/dom';
+
+const WL_ANCHOR_ATTR = [
+  'href',
+  'id',
+  'rel',
+  'rev',
+];
+const PREFIX_DATA_ATTR = /^vars(.+)/;
+const REG_DOMAIN_URL = /^https?:\/\/(www\.)?([^\/:]*)(:\d+)?(\/.*)?$/;
+const PAGE_PROP_WHITELIST = {
+  'SOURCE_URL': true,
+  'DOCUMENT_REFERRER': true,
+};
+
+export class LinkRewriter {
+  /**
+   * @param {!AmpElement} ampElement
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampDoc
+   */
+  constructor(ampElement, ampDoc) {
+    /** @private {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = ampDoc;
+
+    /** @private {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    /** @private {?Event} */
+    this.event_ = null;
+
+    /** @private {?Object} */
+    this.configOpts_ = getConfigOpts(ampElement);
+
+    /** @public {string} */
+    this.rewrittenUrl = this.configOpts_.output;
+
+    /** @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacementService_ = Services.urlReplacementsForDoc(ampElement);
+  }
+
+  /**
+   * @param {!Event} event
+   */
+  handleClick(event) {
+    this.event_ = event;
+
+    const htmlElement = this.event_.srcElement;
+    const sourceTrimmedDomain = Services
+        .documentInfoForDoc(this.ampDoc_)
+        .sourceUrl.replace(/(www\.)?(.*)/, '$2');
+    const canonicalTrimmedDomain = Services
+        .documentInfoForDoc(this.ampDoc_)
+        .sourceUrl.replace(/(www\.)?(.*)/, '$2');
+
+    if (!htmlElement) {
+      return;
+    }
+
+    if (this.isInternalLink_(htmlElement, sourceTrimmedDomain)
```

Comment:
nit: Can we have isInternalLink_ accept an array. And so we don't need to run the `REG_DOMAIN_URL.test(href)` test twice? Thanks